tests/kafkatest/utils/util.py

```diff
@@ -155,7 +155,13 @@ def validate_delivery(acked, consumed, idempotence_enabled=False, check_lost_dat
         if check_lost_data:
             to_validate = list(missing)[0:1000 if len(missing) > 1000 else len(missing)]
             data_lost = check_lost_data(to_validate)
-            msg = annotate_data_lost(data_lost, msg, len(to_validate))
+
+            # With older versions of message format before KIP-101, data loss could occur due to truncation.
+            # These records won't be in the data logs. Tolerate limited data loss for this case.
+            if len(data_lost) == len(to_validate) and len(missing) <= data_loss_tolerance:
```

Comment:
What does the first condition (`len(data_lost) == len(to_validate)`) achieve?

Reply:
@ijuma Thanks for the review. The specific case I was trying to fix was when the log was truncated and hence acked records not received by the consumer are not in the log. The first condition checks that. We will fail the test if consumer doesn't receive records that are present in the log. Does that make sense?

Reply:
Would we want to check all of `missing` and not just the `to_validate` subset in this case? Also, it's a bit confusing to set `success` to `True`, then to `False` and then set it back to `True` again. Can we refactor the logic a bit?

Reply:
@ijuma Updated the code to make it more obvious that we are checking all of missing and not a subset.